src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenCallTests.cs

```diff
@@ -28674,6 +28674,2028 @@ .maxstack  4
+  IL_0012:  brtrue.s   IL_001f
+  IL_0014:  ldloc.2
+  IL_0015:  ldobj      ""T""
+  IL_001a:  stloc.1
+  IL_001b:  ldloca.s   V_1
+  IL_001d:  br.s       IL_0020
+  IL_001f:  ldloc.2
+  IL_0020:  stloc.0
+  IL_0021:  ldarga.s   V_0
+  IL_0023:  call       ""int Program.GetOffset<T>(ref T)""
+  IL_0028:  stloc.3
+  IL_0029:  ldloc.0
+  IL_002a:  ldc.i4.0
+  IL_002b:  ldloc.0
+  IL_002c:  constrained. ""T""
+  IL_0032:  callvirt   ""int IMoveable.Length.get""
+  IL_0037:  ldloc.3
+  IL_0038:  sub
+  IL_0039:  constrained. ""T""
+  IL_003f:  callvirt   ""IMoveable IMoveable.Slice(int, int)""
+  IL_0044:  pop
+  IL_0045:  ret
+}
+");
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        [WorkItem(63221, "https://github.com/dotnet/roslyn/issues/63221")]
+        public void GenericTypeParameterAsReceiver_ImpicitRangeIndexer_RangeExpression_Struct()
+        {
+            var source = @"
+using System;
+
+interface IMoveable
+{
+    IMoveable Slice(int start, int length);
+    int Length {get;}
+}
+
+struct Item : IMoveable
+{
+    public string Name {get; set;}
+
+    public IMoveable Slice(int start, int length)
+    {
+        Console.WriteLine(""Position Slice for item '{0}'"", Name);
+        return this;
+    }
+
+    public int Length
+    {
+        get
+        {
+            Console.WriteLine(""Position Length for item '{0}'"", Name);
+            return 10;
+        }
+    }
+}
+
+class Program
+{
+    static void Main()
+    {
+        var item1 = new Item {Name = ""1""};
+        Shift1(item1);
+
+        var item2 = new Item {Name = ""2""};
+        Shift2(item2);
+    }
+
+    static void Shift1<T>(T item) where T : struct, IMoveable
+    {
+        _ = item[0..^GetOffset(ref item)];
+    }
+
+    static void Shift2<T>(T item) where T : IMoveable
+    {
+        _ = item[0..^GetOffset(ref item)];
+    }
+    
+    static int value = 0;
+    static int GetOffset<T>(ref T item)
+    {
+        item = (T)(IMoveable)new Item {Name = (--value).ToString()};
+        return 0;
+    }
+}
+";
+
+            var verifier = CompileAndVerify(source, targetFramework: TargetFramework.NetLatest, options: TestOptions.ReleaseExe, expectedOutput: @"
+Position Length for item '-1'
+Position Slice for item '-1'
+Position Length for item '-2'
+Position Slice for item '-2'
+").VerifyDiagnostics();
+
+            verifier.VerifyIL("Program.Shift1<T>",
+@"
+{
+  // Code size       40 (0x28)
+  .maxstack  4
+  .locals init (T& V_0,
+                int V_1)
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  stloc.0
+  IL_0003:  ldarga.s   V_0
+  IL_0005:  call       ""int Program.GetOffset<T>(ref T)""
+  IL_000a:  stloc.1
+  IL_000b:  ldloc.0
+  IL_000c:  ldc.i4.0
+  IL_000d:  ldloc.0
+  IL_000e:  constrained. ""T""
+  IL_0014:  callvirt   ""int IMoveable.Length.get""
+  IL_0019:  ldloc.1
+  IL_001a:  sub
+  IL_001b:  constrained. ""T""
+  IL_0021:  callvirt   ""IMoveable IMoveable.Slice(int, int)""
+  IL_0026:  pop
+  IL_0027:  ret
+}
+");
```

Comment:
Should we include `VerifyIL("Shift2<T>", ...);` as well here, and in `GenericTypeParameterAsReceiver_ImpicitRangeIndexer_RangeExpression_Struct_Ref`?

Reply:
> Should we include ```VerifyIL("Shift2<T>", ...);``` as well here, and in ...

That method is the same as in the "Class" flavor of the test and it's IL is verified there.

Reply:
Here we only need to verify its runtime behavior with an input which is a struct